File: Python/백준/Silver/1991. 트리 순회/트리 순회.py
# node class 정의
class Node:
    # 사실상 자식 없어도 자식 이름 '.'으로 들어와서 기본값 설정 안 해도 됨
    # 자식들 링크가 아닌 이름 저장 (딕셔너리로 찾아다닐 거라 이렇게 설정)
    def __init__(self, name, left=None, right=None):
        self.name = name
        self.left = left
        self.right = right

# 함수들은 각각 루트노드의 이름, 노드 정보 담긴 딕셔너리를 매개변수로 받음
# 전위 순회 함수
def preorder(nodeName, nodeDict):
    if(nodeName == '.'):
        return
    print(nodeName, end='')
    preorder(nodeDict[nodeName].left, nodeDict)
    preorder(nodeDict[nodeName].right, nodeDict)

# 중위 순회 함수
def inorder(nodeName, nodeDict):
    if(nodeName == '.'):
        return
    inorder(nodeDict[nodeName].left, nodeDict)
    print(nodeName, end='')
    inorder(nodeDict[nodeName].right, nodeDict)

# 후위 순회 함수
def postorder(nodeName, nodeDict):
    if(nodeName == '.'):
        return
    postorder(nodeDict[nodeName].left, nodeDict)
    postorder(nodeDict[nodeName].right, nodeDict)
    print(nodeName, end='')


# 노드 개수 입력 받기
n = int(input())

# key: node's name
# value : Node 객체
# 위의 구성을 가지는 딕셔너리 생성
nodeDict = dict()
for _ in range(n):
    names = input().split()
    # 노드가 존재하지 않는 경우
    if(nodeDict.get(names[0]) is None):
        # 노드를 생성하고 저장하도록 함
        nodeDict[names[0]] = Node(names[0], names[1], names[2])

# 함수 사용하여 탐색 진행
preorder('A', nodeDict)
print()
inorder('A', nodeDict)
print()
postorder('A', nodeDict)


    

File: Java/백준/Silver/1991. 트리 순회/트리 순회.java

import org.w3c.dom.Node;

import java.io.*;
import java.util.HashMap;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        HashMap<String, Node> hasNode = new HashMap<>();

        for(int i=0; i<n; i++){
            String input[] = br.readLine().split(" ");
            String parent = input[0];
            hasNode.computeIfAbsent(parent, k -> new Node(parent));
            Node left = new Node(input[1]);
            Node right = new Node(input[2]);
            hasNode.get(parent).left = left;
            hasNode.get(parent).right = right;
            hasNode.put(input[1], left);
            hasNode.put(input[2], right);
        }
        
        preorder(hasNode.get("A"));
        bw.write("\n");
        inorder(hasNode.get("A"));
        bw.write("\n");
        postorder(hasNode.get("A"));
        bw.close();
    }

    static void preorder(Node root) throws IOException {
        if(root.name.equals(".")) return;
        bw.write(root.name);
        preorder(root.left);
        preorder(root.right);
    }

    static void inorder(Node root) throws IOException {
        if(root.name.equals(".")) return;
        inorder(root.left);
        bw.write(root.name);
        inorder(root.right);
    }

    static void postorder(Node root) throws IOException {
        if(root.name.equals(".")) return;
        postorder(root.left);
        postorder(root.right);
        bw.write(root.name);
    }

    static class Node{
        String name;
        Node left;
        Node right;
        Node(String name){
            this.name = name;
        }
    }
}